lib/dai-plugin-mcd/src/QueryApiMcdService.js

```diff
@@ -0,0 +1,142 @@
+import { PublicService } from '@makerdao/services-core';
+import fetch from 'isomorphic-fetch';
+import { ServiceRoles } from './constants';
+import { RAY } from './constants';
+import { MDAI } from './index';
+import BigNumber from 'bignumber.js';
+import { stringToBytes } from './utils';
+import { TESTNET_ID } from '../../../contracts/networks';
+
+const LOCAL_URL = 'http://localhost:5000/graphql';
+const KOVAN_SERVER_URL = 'http://vdb0-testing.vulcanize.io:5000/graphql';
+
+export default class QueryApi extends PublicService {
+  constructor(name = ServiceRoles.QUERY_API) {
+    super(name, ['web3']);
+  }
+
+  connect() {
+    const network = this.get('web3').network;
+    switch (network) {
+      case TESTNET_ID:
+        this.serverUrl = LOCAL_URL;
+        break;
+      case 'kovan':
+      case 42:
+      default:
+        this.serverUrl = KOVAN_SERVER_URL;
+        break;
+    }
+  }
+
+  async _blockNumberToDateTime(blockNumber) {
+    const timestamp = (await this.get('web3').getBlock(blockNumber)).timestamp;
+    return new Date(1000 * timestamp);
+  }
+
+  //TODO: add error handling, and handle no return value from vulcanize?
+  async getCdpEventsForIlkAndUrn(
+    ilk,
+    urn,
+    currency,
+    ilkNameReplacement = null
+  ) {
+    let ilkName = stringToBytes(ilk);
+    if (ilkNameReplacement) ilkName = ilkNameReplacement; //only used for testing when the ilks in vdb don't match the ilks in the sdk
+    //todo: confirm whether it really is the case that the ilks don't match?
+    const query = `query ($ilkName: String!, $urn: String!){
+      urnFrobs(_ilkName: $ilkName, _urn: $urn) {
+        nodes {
+          dart
+          dink
+          ilk {
+            nodes {
+              rate
+            }
+          }
+          tx {
+            transactionHash
+            txFrom
+            era {
+              iso
+            }
+          }
+          urn {
+            nodes {
+              art
+              ink
+            }
+          }
+        }
+      }
+    }`;
+    const resp = await fetch(this.serverUrl, {
+      method: 'POST',
+      headers: {
+        Accept: 'application/json',
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify({
+        query,
+        variables: { ilkName, urn }
+      })
+    });
+
+    const { data } = await resp.json();
+    return Promise.all(
+      data.urnFrobs.nodes.map(async e => {
```

Comment:
what you're doing in this nested method is basically transforming an event so it becomes human-readable. this would be a nice thing to pull out into a method at the top-level of the file, and then export it so that it can be unit tested directly, without involving network round-trips or APIs, e.g.:

```
const rawEvent = { ... };
const expected = { ... };
const formattedEvent = formatFrob(rawEvent); // or whatever name
expect(formattedEvent).toEqual(expected);
```